### PR TITLE
Desktop app can compile and load into programming module

### PR DESCRIPTION
**Pull request by baixiaobest**

### CODE REVIEW

File: arduino_files/test_led/LEDArrayConnection.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+#ifndef SRC_LEDARRAYCONNECTION_H_
+#define SRC_LEDARRAYCONNECTION_H_
+
+#include <Arduino.h>
+
+class LEDArrayConnection {
+private:
+	HardwareSerial* m_serial;
+	char m_LEDState;
```
Comment (baiyu123) [inline arduino_files/test_led/LEDArrayConnection.h:20]: Add a documentation on this

File: arduino_files/test_led/Message.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/*
+ * ParsedMessage.h
```
Comment (baiyu123) [inline arduino_files/test_led/Message.h:2]: need documentation, where and how to use this class.

File: arduino_files/test_led/MessageParser.cpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+    return parseMessageBodyAndBuildMessage(m_message + headerEndIdx + 1, dataType);
+}
+
+Message MessageParser::getParsedMessage() {
+	return m_parsedMessage;
+}
+
+//bool MessageParser::isCorrupted() {
```
Comment (baiyu123) [inline arduino_files/test_led/MessageParser.cpp:46]: Delete this function

File: arduino_files/test_led/SerializerUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    buffer[1] = '\0';
+//    DataType dataType = getDataType<T>(input);
+    
+    strcat(buffer, DATA_TYPE_STRINGS[dataType]);
+    strncat(buffer, &HEADER_SEPARATOR,1);
+    char dataString[DATA_STRING_SIZE];
+    DataToString(dataString, input, dataType);
+    
+    strncat(buffer, dataString, MAX_HEADER_SIZE);
+//    strncat(buffer, &ESCAPE_CHAR,1);
```
Comment (baiyu123) [inline arduino_files/test_led/SerializerUtils.cpp]: delete it

File: arduino_files/test_led/SerializerUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            break;
+    }
+}
+
+template <typename T>
+void Serialize(char* buffer, T input, DataType dataType)
+{
+    buffer[0] = START_CHAR;
+    buffer[1] = '\0';
+//    DataType dataType = getDataType<T>(input);
```
Comment (baiyu123) [inline arduino_files/test_led/SerializerUtils.cpp]: delete

File: arduino_files/test_led/SerializerUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+void SerializeString(char* buffer, char* input)
+{
+    buffer[0] = START_CHAR;
+    buffer[1] = '\0';
+    strncat(buffer, DATA_TYPE_STRINGS[DataType::STRING], MAX_HEADER_SIZE);
+    strncat(buffer, &HEADER_SEPARATOR,1);
+    strncat(buffer, input, MAX_MESSAGE_BODY_SIZE);
+//    strncat(buffer, &ESCAPE_CHAR,1);
```
Comment (baiyu123) [inline arduino_files/test_led/SerializerUtils.cpp]: delete

File: arduino_files/test_led/SerializerUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    while(*str != '\0')
+    {
+        checksum ^= *str;
+        str++;
+    }
+    return checksum;
+}
+
+//template<typename T>
+//DataType StringVariable::getDataType(T &input)
```
Comment (baiyu123) [inline arduino_files/test_led/SerializerUtils.cpp]: delete

File: arduino_files/test_led/SerializerUtils.h [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+#ifndef STRINGVARIABLE_H_
+#define STRINGVARIABLE_H_
+#include "SerializerUtils.h"
+#include "MessageConstants.h"
+#include "stdlib.h"
+#include "stdio.h"
+#include "string.h"
+//#include <typeinfo>
```
Comment (baiyu123) [inline arduino_files/test_led/SerializerUtils.h]: delete

File: arduino_files/test_led/SerializerUtils.h [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+//input a variable and it will serialize it
+template <typename T>
+void Serialize(char* buffer, T input, sp::DataType dataType);
+void SerializeString(char* buffer, char* input);
+
+//checksum
+static char calculateCheckSum(char* str);
+
+    //get data type
+//    template<typename T>
```
Comment (baiyu123) [inline arduino_files/test_led/SerializerUtils.h]: delete

**Review by baiyu123 — CHANGES_REQUESTED** [review]
Good

File: arduino_files/test_led/LEDArrayConnection.h (hunk 1 above)
Comment (baixiaobest) [reply arduino_files/test_led/LEDArrayConnection.h:20]: Comment is on the top of this file.

**Review by baiyu123 — APPROVED** [review]

**Pull request merged**